tools/generate_events.py

```diff
@@ -0,0 +1,284 @@
+
+func New{{eventName}}(
+{%- for p, t in event.properties.items() %}
+    {{event.camel_case(p)}} {{event.go_type(t)}},
+{%- endfor %}
+) *{{eventName}} {
+    return &{{eventName}} {
+{%- for p, t in event.properties.items() %}
+        {{event.pascal_case(p)}}: {{event.camel_case(p)}},
+{%- endfor %}
+    }
+}
+    
+func (e *{{eventName}}) GetId() string {
+    return {{eventName}}_Id
+}
+
+func (e *{{eventName}}) GetSeverity() string {
+    return "{{event.event_severity()}}"
+}
+
+func (e *{{eventName}}) GetClusterId() *strfmt.UUID {
+    return &e.ClusterId
+}
+
+{%- if event.type != "cluster" %}
+func (e *{{eventName}}) GetHostId() *strfmt.UUID {
+    return &e.HostId
+}
+{%- endif %}
+
+func (e *{{eventName}}) format(message *string) string {
+    r := strings.NewReplacer(
+{%- for p, t in event['properties'].items() %}
+    "{{'{'}}{{p}}{{'}'}}", fmt.Sprint(e.{{event.pascal_case(p)}}),
+{%- endfor %}
+                  )
+          return r.Replace(*message)
+}
+
+func (e *{{eventName}}) FormatMessage() string {
+    s := "{{event.format}}"
+    return e.format(&s)
+}
+
+func (e *{{eventName}}) FormatLogMessage() string {
+{% if not event.log_format %}
+    return e.FormatMessage()
+{% else %}
+    s := "{{event.log_format}}"
+    return e.format(&s)
+{% endif %}
+}
+
+{% endfor -%}
+
+''')
+
+
+class EventGenerator(object):
+    VALID_PROPS_TYPES = {"UUID":dict(go_type="strfmt.UUID", go_import="github.com/go-openapi/strfmt"),
+                         "integer":dict(go_type="int", go_import=None),
+                         "string":dict(go_type="string", go_import=None)}
+    id_counter = 0
+    def __init__(self, event):
+        self.event = event
+
+    def __getattr__(self, attr):
+        #import ipdb; ipdb.set_trace()
+        if attr in self.__dict__:
+            return getattr(self, attr)
+        else:
+            return getattr(self.event, attr)
+
+    def ctor_name(self):
+        return "New" + self.event_class()
+
+    def ctor_args(self):
+        args = []
+        for arg, _ in self.event.properties.iteritems():
+            args.append(EventGenerator.camel_case(arg) + " " + EventGenerator.pascal_case(arg))
+        return ", ".join(args)
+
+    def event_class(self):
+        return EventGenerator.pascal_case(self.event.name) + "Event"
+
+    def event_type(self):
+        return "Type" + EventGenerator.pascal_case(self.event.name)
+
+    def event_severity(self):
+        #if self.event.severity.lower() == "debug":
+        #    return "EventSeverityDebug"
+        if self.event.severity.lower() == "info":
+            return "info"
+        if self.event.severity.lower() == "warning":
+            return "warning"
+        if self.event.severity.lower() == "error":
+            return "error"
+        if self.event.severity.lower() == "critical":
+            return "critical"
+
+    def get_id(self):
+        EventGenerator.id_counter += 1
+        return "%s%04d" % (self.event.id_prefix, self.id_counter)
```

Comment:
This means we need to be careful to always append to the file right?

Reply:
You are right. 
Didn;t think of it.
I think it is better to have the  full ID in the events.yaml and just verify that there are no duplicates, in the generation tool.
What do you think?

Reply:
fixed

Reply:
What is the ID for?  It still seems hard to maintain.  Can we just use name for ID?

Reply:
ID can be whatever we decide.
The code generation script verifies that the ID is unique
I just started some (dum) convention for the ID names, but we can decide on something that makes sense. Only requirement is that it is unique